File: Samples/WiFiScan/cpp/DisplayHelpers.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include <ppltasks.h>
#include "DisplayHelpers.h"
#include "MainPage.xaml.h"

using namespace concurrency;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Devices::WiFi;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Networking::Connectivity;
using namespace Windows::UI::Core;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Media::Imaging;

using namespace SDKTemplate::WiFiScan;
namespace SDKTemplate {
    namespace WiFiScan {

        // WiFiNetworkDisplay

        WiFiNetworkDisplay::WiFiNetworkDisplay(WiFiAvailableNetwork^ availableNetwork, WiFiAdapter^ adapter)
        {
            AvailableNetwork = availableNetwork;
            this->adapter = adapter;
            UpdateWiFiImage();
        }

        void WiFiNetworkDisplay::UpdateWiFiImage()
        {
            Platform::String^ imageFileNamePrefix;
            if (AvailableNetwork->SecuritySettings->NetworkAuthenticationType == NetworkAuthenticationType::Open80211)
            {
                imageFileNamePrefix = L"open";
            }
            else
            {
                imageFileNamePrefix = L"secure";
            }

            Platform::String^ imageFileName = L"ms-appx:/Assets/" + imageFileNamePrefix + "_" + AvailableNetwork->SignalBars + "bar.png";

            WiFiImage = ref new BitmapImage(ref new Uri(imageFileName));

            OnPropertyChanged("WiFiImage");
        }

        task<void> WiFiNetworkDisplay::UpdateConnectivityLevels(IObservableVector<WiFiNetworkDisplay^>^ networkCollection, UINT current)
        {
            if (current < networkCollection->Size)
            {
                try {
                    return networkCollection->GetAt(current)->UpdateConnectivityLevel().then([networkCollection, current]()
                    {
                        // Wait until the first task is done, then recursively call the next one
                        // This way we won't call GetConnectedProfileAsync() too many times 
                        // which may exhaust the available wlan handles
                        return UpdateConnectivityLevels(networkCollection, current + 1);
                    });
                }
                catch (...)
                {
                    // The network collection may have changed. If that happens,
                    // return an empty task to terminate the recusrsion
                    return create_task([]()
                    {
                    });
                }
            }
            else
            {
                return create_task([]()
                {
                });
            }
        }

        task<void> WiFiNetworkDisplay::UpdateConnectivityLevel()
        {
            auto connectionProfileOperation = adapter->NetworkAdapter->GetConnectedProfileAsync();
            auto connectionProfileTask = create_task(connectionProfileOperation);

            return connectionProfileTask.then([this](ConnectionProfile^ connectedProfile)
            {
                Platform::String^ connectedSsid = nullptr;

                if (connectedProfile != nullptr &&
                    connectedProfile->IsWlanConnectionProfile &&
                    connectedProfile->WlanConnectionProfileDetails != nullptr)
                {
                    connectedSsid = connectedProfile->WlanConnectionProfileDetails->GetConnectedSsid();
                }

                if (connectedSsid != nullptr &&
                    AvailableNetwork->Ssid != nullptr &&
                    connectedSsid->Equals(AvailableNetwork->Ssid))
                {
                    ConnectivityLevel = WiFiNetworkDisplay::GetConnectivityLevelAsString(connectedProfile->GetNetworkConnectivityLevel());
                }
                else
                {
                    ConnectivityLevel = L"Not Connected";
                }

                OnPropertyChanged("ConnectivityLevel");
            });
        }

        Platform::String^ WiFiNetworkDisplay::GetNetworkAuthenticationTypeAsString(NetworkAuthenticationType authenticationType)
        {
            switch (authenticationType)
            {
            case NetworkAuthenticationType::None:
                return L"None";
            case NetworkAuthenticationType::Unknown:
                return L"Unknown";
            case NetworkAuthenticationType::Open80211:
                return L"Open80211";
            case NetworkAuthenticationType::SharedKey80211:
                return L"SharedKey80211";
            case NetworkAuthenticationType::Wpa:
                return L"Wpa";
            case NetworkAuthenticationType::WpaPsk:
                return L"WpaPsk";
            case NetworkAuthenticationType::WpaNone:
                return L"WpaNone";
            case NetworkAuthenticationType::Rsna:
                return L"Rsna";
            case NetworkAuthenticationType::RsnaPsk:
                return L"RsnaPsk";
            default:
                return L"Error getting Network Authentication Type";
            }
        }

        Platform::String^ WiFiNetworkDisplay::GetNetworkEncryptionTypeAsString(NetworkEncryptionType encryptionType)
        {
            switch (encryptionType)
            {
            case NetworkEncryptionType::None:
                return L"None";
            case NetworkEncryptionType::Unknown:
                return L"Unknown";
            case NetworkEncryptionType::Wep:
                return L"Wep";
            case NetworkEncryptionType::Wep40:
                return L"Wep40";
            case NetworkEncryptionType::Wep104:
                return L"Wep104";
            case NetworkEncryptionType::Tkip:
                return L"Tkip";
            case NetworkEncryptionType::Ccmp:
                return L"Ccmp";
            case NetworkEncryptionType::WpaUseGroup:
                return L"WpaUseGroup";
            case NetworkEncryptionType::RsnUseGroup:
                return L"RsnUseGroup";
            default:
                return L"Error getting Network Encryption Type";
            }
        }

        Platform::String^ WiFiNetworkDisplay::GetConnectionStatusAsString(WiFiConnectionStatus connectionStatus)
        {
            switch (connectionStatus)
            {
            case WiFiConnectionStatus::UnspecifiedFailure:
                return L"UnspecifiedFailure";
            case WiFiConnectionStatus::Success:
                return L"Success";
            case WiFiConnectionStatus::AccessRevoked:
                return L"AccessRevoked";
            case WiFiConnectionStatus::InvalidCredential:
                return L"InvalidCredential";
            case WiFiConnectionStatus::NetworkNotAvailable:
                return L"NetworkNotAvailable";
            case WiFiConnectionStatus::Timeout:
                return L"Timeout";
            case WiFiConnectionStatus::UnsupportedAuthenticationProtocol:
                return L"UnsupportedAuthenticationProtocol";
            default:
                return L"Error getting Connection Status";
            }
        }

        Platform::String^ WiFiNetworkDisplay::GetConnectivityLevelAsString(NetworkConnectivityLevel conectivityLevel)
        {
            switch (conectivityLevel)
            {
            case NetworkConnectivityLevel::ConstrainedInternetAccess:
                return L"ConstrainedInternetAccess";
            case NetworkConnectivityLevel::InternetAccess:
                return L"InternetAccess";
            case NetworkConnectivityLevel::LocalAccess:
                return  L"LocalAccess";
            case NetworkConnectivityLevel::None:
                return  L"Not Connected";
            default:
                return L"Not Connected";
            }
        }

        void WiFiNetworkDisplay::OnPropertyChanged(String^ name)
        {
            MainPage::Current->Dispatcher->RunAsync(CoreDispatcherPriority::Normal, ref new DispatchedHandler([=]()
            {
                PropertyChanged(this, ref new PropertyChangedEventArgs(name));
            }));
        }
    }
}

File: Samples/WiFiScan/cs/DisplayHelpers.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using SDKTemplate;
using System;
using System.ComponentModel;
using System.Threading.Tasks;
using Windows.Devices.WiFi;
using Windows.Networking.Connectivity;
using Windows.UI.Core;
using Windows.UI.Xaml.Media.Imaging;

namespace WiFiScan
{
    public class WiFiNetworkDisplay : INotifyPropertyChanged
    {
        private WiFiAdapter adapter;
        public WiFiNetworkDisplay(WiFiAvailableNetwork availableNetwork, WiFiAdapter adapter)
        {
            AvailableNetwork = availableNetwork;
            this.adapter = adapter;
            UpdateWiFiImage();
        }

        private void UpdateWiFiImage()
        {
            string imageFileNamePrefix = "secure";
            if (AvailableNetwork.SecuritySettings.NetworkAuthenticationType == NetworkAuthenticationType.Open80211)
            {
                imageFileNamePrefix = "open";
            }

            string imageFileName = string.Format("ms-appx:/Assets/{0}_{1}bar.png", imageFileNamePrefix, AvailableNetwork.SignalBars);

            WiFiImage = new BitmapImage(new Uri(imageFileName));

            OnPropertyChanged("WiFiImage");

        }

        public async Task UpdateConnectivityLevel()
        {
            string connectivityLevel = "Not Connected";
            string connectedSsid = null;

            var connectedProfile = await adapter.NetworkAdapter.GetConnectedProfileAsync();
            if (connectedProfile != null &&
                connectedProfile.IsWlanConnectionProfile &&
                connectedProfile.WlanConnectionProfileDetails != null)
            {
                connectedSsid = connectedProfile.WlanConnectionProfileDetails.GetConnectedSsid();
            }

            if (!string.IsNullOrEmpty(connectedSsid))
            {
                if (connectedSsid.Equals(AvailableNetwork.Ssid))
                {
                    connectivityLevel = connectedProfile.GetNetworkConnectivityLevel().ToString();
                }
            }

            ConnectivityLevel = connectivityLevel;

            OnPropertyChanged("ConnectivityLevel");
        }

        public String Ssid
        {
            get
            {
                return availableNetwork.Ssid;
            }
        }

        public String Bssid
        {
            get
            {
                return availableNetwork.Bssid;

            }
        }

        public String ChannelCenterFrequency
        {
            get
            {
                return string.Format("{0}kHz", availableNetwork.ChannelCenterFrequencyInKilohertz);
            }
        }

        public String Rssi
        {
            get
            {
                return string.Format("{0}dBm", availableNetwork.NetworkRssiInDecibelMilliwatts);
            }
        }

        public String SecuritySettings
        {
            get
            {
                return string.Format("Authentication: {0}; Encryption: {1}", availableNetwork.SecuritySettings.NetworkAuthenticationType, availableNetwork.SecuritySettings.NetworkEncryptionType);
            }
        }
        public String ConnectivityLevel
        {
            get;
            private set;
        }

        public BitmapImage WiFiImage
        {
            get;
            private set;
        }


        private WiFiAvailableNetwork availableNetwork;
        public WiFiAvailableNetwork AvailableNetwork
        {
            get
            {
                return availableNetwork;
            }

            private set
            {
                availableNetwork = value;
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;
        protected async void OnPropertyChanged(string name)
        {
            var rootPage = MainPage.Current;
            await rootPage.Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
            {
                PropertyChangedEventHandler handler = PropertyChanged;
                if (handler != null)
                {
                    handler(this, new PropertyChangedEventArgs(name));
                }
            });
        }
    }
}
